File: scala-client/src/main/scala/com/couchbase/client/scala/view/ViewOptions.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.scala.view

import com.couchbase.client.core.annotation.Stability.Volatile
import com.couchbase.client.core.cnc.RequestSpan
import com.couchbase.client.core.retry.RetryStrategy
import com.couchbase.client.scala.json.JsonArray
import com.couchbase.client.scala.transformers.JacksonTransformers

import scala.concurrent.duration.Duration
import scala.util.Try

/** Customize the execution of a view.
  *
  * @author Graham Pople
  * @since 1.0.0
  */
case class ViewOptions(
    private[scala] val namespace: Option[DesignDocumentNamespace] = None,
    private[scala] val reduce: Option[Boolean] = None,
    private[scala] val limit: Option[Int] = None,
    private[scala] val group: Option[Boolean] = None,
    private[scala] val groupLevel: Option[Int] = None,
    private[scala] val inclusiveEnd: Option[Boolean] = None,
    private[scala] val skip: Option[Int] = None,
    private[scala] val onError: Option[ViewErrorMode] = None,
    private[scala] val debug: Option[Boolean] = None,
    private[scala] val order: Option[ViewOrdering] = None,
    private[scala] val key: Option[String] = None,
    private[scala] val startKeyDocId: Option[String] = None,
    private[scala] val endKeyDocId: Option[String] = None,
    private[scala] val endKey: Option[String] = None,
    private[scala] val startKey: Option[String] = None,
    private[scala] val keys: Option[String] = None,
    private[scala] val retryStrategy: Option[RetryStrategy] = None,
    private[scala] val timeout: Option[Duration] = None,
    private[scala] val scanConsistency: Option[ViewScanConsistency] = None,
    private[scala] val parentSpan: Option[RequestSpan] = None
) {

  /** Sets the parent `RequestSpan`.
    *
    * @return a copy of this with the change applied, for chaining.
    */
  @Volatile
  def parentSpan(value: RequestSpan): ViewOptions = {
    copy(parentSpan = Some(value))
  }

  /** Sets in which namespace the view exists.
    *
    * @return this for further chaining
    */
  def namespace(value: DesignDocumentNamespace): ViewOptions = {
    copy(namespace = Some(value))
  }

  /** Explicitly enable/disable the reduce function on the query.
    *
    * @return this for further chaining
    */
  def reduce(value: Boolean): ViewOptions = {
    copy(reduce = Some(value))
  }

  /** Limit the number of the returned documents to the specified number.
    *
    * @return this for further chaining
    */
  def limit(value: Int): ViewOptions = {
    copy(limit = Some(value))
  }

  /** Group the results using the reduce function to a group or single row.
    *
    * Important: this setter and `groupLevel` should not be used
    * together. It is sufficient to only set the
    * grouping level only and use this setter in cases where you always want the
    * highest group level implicitly.
    *
    * @return this for further chaining
    */
  def group(value: Boolean): ViewOptions = {
    copy(group = Some(value))
  }

  /** Specifies the depth within the key to group results.
    *
    * Important: this setter and `groupLevel` should not be used
    * together. It is sufficient to only set the
    * grouping level only and use this setter in cases where you always want the
    * highest group level implicitly.
    *
    * @return this for further chaining
    */
  def groupLevel(level: Int): ViewOptions = {
    copy(groupLevel = Some(level))
  }

  /** Specifies whether the specified end key should be included in the result.
    *
    * @return this for further chaining
    */
  def inclusiveEnd(value: Boolean): ViewOptions = {
    copy(inclusiveEnd = Some(value))
  }

  /** Skip this number of records before starting to return the results.
    *
    * @return this for further chaining
    */
  def skip(value: Int): ViewOptions = {
    copy(skip = Some(value))
  }

  /** Sets the response in the event of an error.
    *
    * @return this for further chaining
    */
  def onError(value: ViewErrorMode): ViewOptions = {
    copy(onError = Some(value))
  }

  /** Enable debugging on view queries.
    *
    * @return this for further chaining
    */
  def debug(value: Boolean): ViewOptions = {
    copy(debug = Some(value))
  }

  /** Specifies the results ordering.  See [[ViewOrdering]] for details.
    *
    * @return this for further chaining
    */
  def order(value: ViewOrdering): ViewOptions = {
    copy(order = Some(value))
  }

  /** Specifies a specific key to fetch from the index.
    *
    * @return this for further chaining
    */
  def key(value: Any): ViewOptions = {
    value match {
      case s: String => copy(key = Some("\"" + s + "\""))
      case _         => copy(key = Some(value.toString))
    }
  }

  /** Specifies the document id to start returning results at within a number of results should `startKey` have multiple
    * entries within the index.
    *
    * @return this for further chaining
    */
  def startKeyDocId(value: String): ViewOptions = {
    copy(startKeyDocId = Some(value))
  }

  /** Specifies the document id to stop returning results at within a number of results should `endKey` have multiple
    * entries within the index.
    *
    * @return this for further chaining
    */
  def endKeyDocId(value: String): ViewOptions = {
    copy(endKeyDocId = Some(value))
  }

  /** Specifies the key to stop returning results at.
    *
    * @return this for further chaining
    */
  def endKey(value: Any): ViewOptions = {
    value match {
      case s: String =>
        copy(endKey = Try(JacksonTransformers.MAPPER.writeValueAsString(value)).toOption)
      case _ => copy(endKey = Some(value.toString))
    }
  }

  /** Specifies the key to skip to before beginning to return results.
    *
    * @return this for further chaining
    */
  def startKey(value: Any): ViewOptions = {
    value match {
      case s: String =>
        copy(startKey = Try(JacksonTransformers.MAPPER.writeValueAsString(value)).toOption)
      case _ => copy(startKey = Some(value.toString))
    }
  }

  /** Specifies a specific set of keys to fetch from the index.
    *
    * @return this for further chaining
    */
  def keys(values: Iterable[Any]): ViewOptions = {
    val arr = JsonArray.create
    values.foreach(v => {
      arr.add(v match {
        case x: String => "\"" + x + "\""
        case _         => v.toString
      })
    })
    copy(keys = Some(arr.toString))
  }

  /** When to timeout the operation.
    *
    * @return this for further chaining
    */
  def timeout(timeout: Duration): ViewOptions = {
    copy(timeout = Option(timeout))
  }

  /** Sets what retry strategy to use if the operation fails.
    *
    * @param strategy the retry strategy to use
    *
    * @return a copy of this with the change applied, for chaining.
    */
  def retryStrategy(strategy: RetryStrategy): ViewOptions = {
    copy(retryStrategy = Some(strategy))
  }

  /** Sets what scan consistency to use.  See [[ViewScanConsistency]] for details.
    *
    * @param scanConsistency the scan consistency to use
    *
    * @return a copy of this with the change applied, for chaining.
    */
  def scanConsistency(scanConsistency: ViewScanConsistency): ViewOptions = {
    copy(scanConsistency = Some(scanConsistency))
  }

  private[scala] def encode() = {
    val sb = new StringBuilder()
    Seq(
      "reduce"         -> reduce,
      "limit"          -> limit,
      "group"          -> group,
      "group_level"    -> groupLevel,
      "inclusive_end"  -> inclusiveEnd,
      "skip"           -> skip,
      "on_error"       -> onError.map(_.encode),
      "debug"          -> debug,
      "descending"     -> order,
      "key"            -> key,
      "startkey_docid" -> startKeyDocId,
      "endkey_docid"   -> endKeyDocId,
      "endkey"         -> endKey,
      "startkey"       -> startKey,
      "stale"          -> scanConsistency.map(_.encoded)
    ).foreach {
      case (key, valueOpt) =>
        valueOpt.foreach(sb ++= key += '=' ++= _.toString += '&')
    }
    sb.toString.stripSuffix("&")
  }
}

File: java-client/src/main/java/com/couchbase/client/java/view/ViewOptions.java
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.view;

import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.error.InvalidArgumentException;
import com.couchbase.client.core.util.UrlQueryStringBuilder;
import com.couchbase.client.java.CommonOptions;
import com.couchbase.client.java.codec.JsonSerializer;
import com.couchbase.client.java.json.JsonArray;
import com.couchbase.client.java.json.JsonObject;

import static com.couchbase.client.core.util.Validators.notNull;

public class ViewOptions extends CommonOptions<ViewOptions> {

  /**
   * Contains all stored params.
   */
  private final UrlQueryStringBuilder params = UrlQueryStringBuilder.createForUrlSafeNames();

  private String keysJson;
  private JsonSerializer serializer;
  private DesignDocumentNamespace namespace;

  public static ViewOptions viewOptions() {
    return new ViewOptions();
  }

  private ViewOptions() {
  }

  public ViewOptions namespace(final DesignDocumentNamespace namespace) {
    this.namespace = namespace;
    return this;
  }

  public ViewOptions serializer(final JsonSerializer serializer) {
    this.serializer = serializer;
    return this;
  }

  /**
   * Explicitly enable/disable the reduce function on the query.
   *
   * @param reduce if reduce should be enabled or not.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions reduce(final boolean reduce) {
    params.set("reduce", reduce);
    return this;
  }


  /**
   * Limit the number of the returned documents to the specified number.
   *
   * @param limit the number of documents to return.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions limit(final int limit) {
    if (limit < 0) {
      throw InvalidArgumentException.fromMessage("Limit must be >= 0.");
    }
    params.set("limit", limit);
    return this;
  }

  /**
   * Group the results using the reduce function to a group or single row.
   * <p>
   * Important: this setter and {@link #groupLevel(int)} should not be used
   * together in the same {@link ViewOptions}. It is sufficient to only set the
   * grouping level only and use this setter in cases where you always want the
   * highest group level implictly.
   *
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions group(boolean group) {
    params.set("group", group);
    return this;
  }

  /**
   * Specify the group level to be used.
   * <p>
   * Important: {@link #group(boolean)} and this setter should not be used
   * together in the same {@link ViewOptions}. It is sufficient to only use this
   * setter and use {@link #group(boolean)} in cases where you always want
   * the highest group level implicitly.
   *
   * @param grouplevel How deep the grouping level should be.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions groupLevel(final int grouplevel) {
    params.set("group_level", grouplevel);
    return this;
  }

  /**
   * Specifies whether the specified end key should be included in the result.
   *
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions inclusiveEnd(boolean inclusive) {
    params.set("inclusive_end", inclusive);
    return this;
  }

  /**
   * Skip this number of records before starting to return the results.
   *
   * @param skip The number of records to skip.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions skip(final int skip) {
    if (skip < 0) {
      throw InvalidArgumentException.fromMessage("Skip must be >= 0.");
    }
    params.set("skip", skip);
    return this;
  }

  /**
   * Sets the scan consistency (staleness) of a view query.
   *
   * @param scanConsistency Which consistency mode should be used.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions scanConsistency(final ViewScanConsistency scanConsistency) {
    params.set("stale", scanConsistency.toString());
    return this;
  }

  /**
   * Sets the response in the event of an error.
   * <p>
   * See the "OnError" enum for more details on the available options.
   *
   * @param viewErrorMode The appropriate error handling type.
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions onError(final ViewErrorMode viewErrorMode) {
    params.set("on_error", viewErrorMode.toString());
    return this;
  }

  /**
   * Enable debugging on view queries.
   *
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions debug(boolean debug) {
    params.set("debug", debug);
    return this;
  }

  /**
   * Return the documents in descending key order.
   *
   * @return the {@link ViewOptions} object for proper chaining.
   */
  public ViewOptions order(final ViewOrdering ordering) {
    notNull(ordering, "ViewOrdering");
    params.set("descending", ordering == ViewOrdering.DESCENDING);
    return this;
  }

  public ViewOptions key(String key) {
    params.set("key", "\"" + key + "\"");
    return this;
  }

  public ViewOptions key(int key) {
    params.set("key", key);
    return this;
  }

  public ViewOptions key(long key) {
    params.set("key", key);
    return this;
  }

  public ViewOptions key(double key) {
    params.set("key", Double.toString(key));
    return this;
  }


  public ViewOptions key(boolean key) {
    params.set("key", key);
    return this;
  }

  public ViewOptions key(JsonObject key) {
    params.set("key", key.toString());
    return this;
  }

  public ViewOptions key(JsonArray key) {
    params.set("key", key.toString());
    return this;
  }

  public ViewOptions keys(JsonArray keys) {
    this.keysJson = keys.toString();
    return this;
  }

  public ViewOptions startKeyDocId(String id) {
    params.set("startkey_docid", id);
    return this;
  }

  public ViewOptions endKeyDocId(String id) {
    params.set("endkey_docid", id);
    return this;
  }

  public ViewOptions endKey(String key) {
    params.set("endkey", "\"" + key + "\"");
    return this;
  }

  public ViewOptions endKey(int key) {
    params.set("endkey", key);
    return this;
  }

  public ViewOptions endKey(long key) {
    params.set("endkey", key);
    return this;
  }

  public ViewOptions endKey(double key) {
    params.set("endkey", String.valueOf(key));
    return this;
  }


  public ViewOptions endKey(boolean key) {
    params.set("endkey", key);
    return this;
  }

  public ViewOptions endKey(JsonObject key) {
    params.set("endkey", key.toString());
    return this;
  }

  public ViewOptions endKey(JsonArray key) {
    params.set("endkey", key.toString());
    return this;
  }

  public ViewOptions startKey(String key) {
    params.set("startkey", "\"" + key + "\"");
    return this;
  }

  public ViewOptions startKey(int key) {
    params.set("startkey", key);
    return this;
  }

  public ViewOptions startKey(long key) {
    params.set("startkey", key);
    return this;
  }

  public ViewOptions startKey(double key) {
    params.set("startkey", String.valueOf(key));
    return this;
  }

  public ViewOptions startKey(boolean key) {
    params.set("startkey", key);
    return this;
  }

  public ViewOptions startKey(JsonObject key) {
    params.set("startkey", key.toString());
    return this;
  }

  public ViewOptions startKey(JsonArray key) {
    params.set("startkey", key.toString());
    return this;
  }

  public ViewOptions raw(String key, String value) {
    params.set(key, value);
    return this;
  }

  /**
   * A string representation of this ViewQuery, suitable for logging and other human consumption.
   * If the {@link #keys(JsonArray)} parameter is too large, it is truncated in this dump.
   * <p>
   * see the {@link #export()} ()} for the parameter representation of the ViewQuery execution URL.
   */
  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("ViewQuery{");
    sb.append("params=\"").append(export()).append('"');
    if (namespace == DesignDocumentNamespace.DEVELOPMENT) {
      sb.append(", dev");
    }
    if (keysJson != null) {
      sb.append(", keys=\"");
      if (keysJson.length() < 140) {
        sb.append(keysJson).append('"');
      } else {
        sb.append(keysJson, 0, 140)
          .append("...\"(")
          .append(keysJson.length())
          .append(" chars total)");
      }
    }
    sb.append('}');
    return sb.toString();
  }

  /**
   * Returns the query string for this ViewQuery, containing all the key/value pairs
   * for parameters that will be part of this ViewQuery's execution URL for the view service.
   */
  String export() {
    return params.build();
  }

  @Stability.Internal
  public Built build() {
    return new Built();
  }

  public class Built extends BuiltCommonOptions {

    Built() { }

    public JsonSerializer serializer() {
      return serializer;
    }

    public String keys() {
      return keysJson;
    }

    public boolean development() {
      return namespace == DesignDocumentNamespace.DEVELOPMENT;
    }

    public String query() {
      return export();
    }

  }

}
